pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -160,6 +220,20 @@ func (rd *restoreDataProcessor) Start(ctx context.Context) {
 	ctx = rd.StartInternal(ctx, restoreDataProcName)
 	rd.input.Start(ctx)
 
+	// First we reserve minWorkerMemReservation for each restore worker, and
+	// making sure that we always have enough memory for at least one worker. The
+	// maximum number of workers is based on the cluster setting. If the cluster
+	// setting is updated, the job should be PAUSEd and RESUMEd for the new
+	// setting to take effect.
+	numWorkers, err := reserveRestoreWorkerMemory(ctx, rd.flowCtx.Cfg.Settings, rd.qp)
+	if err != nil {
+		log.Warningf(ctx, "cannot reserve restore worker memory: %v", err)
+		rd.MoveToDraining(err)
+		return
+	}
```

Comment:
Having the reserve memory happen before the processor does any work is a way to at least make sure the job fails fast if we suspect that there's not enough memory in the node to continue the restore. Restore is something that can take up a lot of cluster resources and I think failing the restore is before it starts is better than failing some heavy SQL query that's in progress. This pattern also exist in our backup job currently where we reserve the worker memory in the beginning, and fail the job if we are not able to do so.

We do have an issue here if the job replans at an unfortunate time and some other queries in the cluster prevents the job from resuming due to lack of memory, but I think in that case it's still better to fail the restore since it wasn't going to be able to continue anyways, and the user can always have an option to pause on error should this become a recurring problem.

Reply:
should this PR point to an issue? I'd like to understand the motivation for this change better. I can imagine 2 different motivations: one is that the current restore frequently causes issues such as OOMing nodes or causing queries to fail, and another is that everything is pretty much ok except for some rare failures but we want to increase the number of workers and therefore we need some memory limits. I guess I have the latter in my mind but maybe I'm wrong.
cc @dt to see what you think.